src/matlab2tikz.m

```diff
@@ -712,19 +714,24 @@ case guitypes()
         % A composite object might nest handleAllChildren calls that can
         % modify the m2t.currentHandleHasLegend value. Re-instate the
         % legend status. For detailed explanations check getLegendEntries().
-        m2t = hasLegendEntry(m2t,child);
-        [m2t,str] = addPlotyyReference(m2t, str, child);
-        str = addLegendInformation(m2t, str, child);
+        m2t               = hasLegendEntry(m2t,child);
+        [m2t, ref1, ref2] = addPlotyyReference(m2t, child);
+        legendInfo        = addLegendInformation(m2t, child);
+        % Add reference 2 BEFORE next plot to preserve color order
+        str               = join(m2t, {ref2, str, ref1, legendInfo}, '');
 
         % append the environment
-        pgfEnvironments{end+1} = str;
+        pgfEnvironments{envCounter} = str;
+        envCounter = envCounter +1;
     end
 end
 % ==============================================================================
-function [m2t, str] = addPlotyyReference(m2t, str, h)
+function [m2t, ref1, ref2] = addPlotyyReference(m2t, h)
```

Comment:
@okomarov Isnt it the other way around 
```matlab
ref1 = sprintf('\\label{plotyyref:leg%d};\n\n', labelNum);

refs{ii}= sprintf('\\addlegendimage{/pgfplots/refstyle=plotyyref:leg%d}\\addlegendentry{%s};\n',...
                              ii, lString);
```
I would have thought, that ref1 is the labelReference and ref2 would be label